OMCompiler/SimulationRuntime/c/CMakeLists.txt

```diff
@@ -1,509 +1,44 @@
-                     COMMAND ${COPY} ${blas_CODE_NEU} ${blas_OUTPUT_NEU}
-                     WORKING_DIRECTORY ${dir}
-                     COMMENT "copy file ${blas_CODE_NEU} to ${blas_OUTPUT_NEU}")
-  # target fuer OM_OUTPUT
-  ADD_CUSTOM_TARGET(blas${model} ALL DEPENDS ${blas_OUTPUT})
-
-  SET(OMC_MODELNAME ${model})
-  # generate model.mos
-
-  # Variablen fuer openmodelica2sarturis
-  SET(OMC_CODE   ${dir}/${model}.c
-                 ${dir}/${model}_functions.c
-                 ${dir}/${model}_init.txt)
-  SET(OMC_OUTPUT  ${dir}/${model}.c
-  #                ${dir}/${model}_functions.cpp
-          ${dir}/${model}_functions.h
-          ${dir}/${model}_records.c)
-  # custom command fuer openmodelicacompiler
-  ADD_CUSTOM_COMMAND(OUTPUT ${OMC_OUTPUT}
-                     COMMAND ${OMC_DEBUG} ${Flags} ${mos}
-                     WORKING_DIRECTORY ${dir}
-                     COMMENT "Generating code for ${model} with ${OMC_DEBUG}")
-  # target fuer OM_OUTPUT
-  ADD_CUSTOM_TARGET(${model}codegen ALL DEPENDS ${OMC_OUTPUT})
-
-  ADD_CUSTOM_TARGET(${model}codegencpp ALL DEPENDS ${OMC_OUTPUT})
-
-  ADD_DEFINITIONS(/TP)
-  set_source_files_properties(${OMC_OUTPUT} PROPERTIES LANGUAGE CXX)
-  ADD_EXECUTABLE(${model} ${OMC_OUTPUT} ${CSRC})
-  TARGET_LINK_LIBRARIES(${model} simulation util math-support results solver meta ModelicaExternalC libexpat initialization lapack_win32_MT ${SUNDIALS_LIBRARIES})
-
-  # custom command to copy xml file
-  SET(XML_CODE   ${dir}/${model}_init.xml)
-  STRING(REGEX REPLACE "/" "\\\\" XML_CODE_NEU ${XML_CODE})
-  SET(XML_OUTPUT  ${CMAKE_CURRENT_BINARY_DIR}/RelWithDebInfo/${model}_init.xml)
-  STRING(REGEX REPLACE "/" "\\\\" XML_OUTPUT_NEU ${XML_OUTPUT})
-  ADD_CUSTOM_COMMAND(OUTPUT ${XML_OUTPUT}
-                     COMMAND ${COPY} ${XML_CODE_NEU} ${XML_OUTPUT_NEU}
-                     WORKING_DIRECTORY ${dir}
-                     COMMENT "copy file ${XML_CODE_NEU} to ${XML_OUTPUT_NEU}")
-  # target fuer OM_OUTPUT
-  ADD_CUSTOM_TARGET(${model}cp_xml ALL DEPENDS ${XML_OUTPUT})
-
-  # Dependencies
-  ADD_DEPENDENCIES(${model}cp_xml ${model}codegen expat${model} lapack${model} blas${model})
-  ADD_DEPENDENCIES(${model} ${model}cp_xml)
-
-#ENDFOREACH(model ${model_sources})
-ENDMACRO(BUILDMODELMOS)
-
-# MACRO BUILDMODEL
-MACRO(BUILDMODELFMU model dir Flags CSRC)
-
-IF(WIN32)
-  SET(COPY copy)
-else(WIN32)
-  SET(COPY cp)
-endif(WIN32)
-
-IF(SUNDIALS_INCLUDE_DIR AND
-   SUNDIALS_LIBRARY_CVODE AND
-   SUNDIALS_LIBRARY_IDA AND
-   SUNDIALS_LIBRARY_NVEC AND
-   SUNDIALS_KINSOL)
-
-  INCLUDE_DIRECTORIES(${SUNDIALS_INCLUDE_DIR})
-  SET(SUNDIALS_LIBRARIES ${SUNDIALS_LIBRARY_CVODE} ${SUNDIALS_LIBRARY_IDA} ${SUNDIALS_LIBRARY_NVEC} ${SUNDIALS_KINSOL})
-
-ENDIF(SUNDIALS_INCLUDE_DIR AND
-      SUNDIALS_LIBRARY_CVODE AND
-      SUNDIALS_LIBRARY_IDA AND
-   SUNDIALS_LIBRARY_NVEC AND
-   SUNDIALS_KINSOL)
-
-  # includes
-  INCLUDE_DIRECTORIES(${OMCTRUNCHOME}/OMCompiler/SimulationRuntime/c)
```

Comment:
How are include directories managed now?

Reply:
This is a good example actually. You now specify include directories separately for each target. 

In this specific case the C source files all use includes relative to their location. So there was no need for an explicit include directory. 

However, libraries that link to these libraries need to know where to find the headers (right?). For those cases, we have this (line 30).

`target_include_directories(OpenModelicaRuntimeC INTERFACE ${CMAKE_CURRENT_SOURCE_DIR})`

This (the `INTERFACE `part) tells every library that links to `OpenModelicaRuntimeC `to add `${CMAKE_CURRENT_SOURCE_DIR}` (which is `SimulationRuntime/c`) to its include directory list, i.e. they will be compiled with `-I  ...SimulationRuntime/c`.

See _OMCompiler/Parser/CMakeLists.txt_ for example. You can see (line 40) that the library _omparse_ is linked to _OpenModelicaRuntimeC_. There is no need to explicitly specify `INCLUDE_DIRECTORIES(${OMCTRUNCHOME}/OMCompiler/SimulationRuntime/c)` here. It will be done automatically because of the interface specifier above.

Reply:
@mahge this is good example of needing to be an expert in CMake to understand all the weird implicit stuff and the need to debug for hours to find a missing include.